drv/opentitan-aontimer/src/lib.rs

```diff
@@ -0,0 +1,154 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+#![no_std]
+
+use core::ptr::slice_from_raw_parts_mut;
+
+const WDOG_REGWEN_IDX: usize = 4;
+const WDOG_CTRL_IDX: usize = 5;
+const WDOG_BARK_THOLD_IDX: usize = 6;
+const WDOG_BITE_THOLD_IDX: usize = 7;
+const WDOG_COUNT_IDX: usize = 8;
+const INTR_STATE_IDX: usize = 9;
+
+const INTR_STATE_BITS_WKUP: u32 = 0x1;
+const INTR_STATE_BITS_WDOG: u32 = 0x2;
+
+const WDOG_CTRL_BITS_EN: u32 = 0x1;
+const WDOG_CTRL_BITS_PAUSE_SLEEP: u32 = 0x2;
+
+// The OpenTitan Always-On Timer is a counter that increments. If the counter
+// reaches a configurable threshold, it will emit an interrupt signal, called the
+// "bark". If the counter reaches a second configurable threshold, it will reset
+// the system, called the "bite".
+// Enabling the AON Timer before setting a bite threshold causes the system to
+// immediately reset.
+pub struct AonTimer {
+    base_addr: *mut [u32],
+    clock_freq_hz: u32,
+    pub bark_cb: Option<fn()>,
+}
+
+impl AonTimer {
+    // Initialize a new watchdog driver device with the specified base address,
+    // clock frequency (in Hz), and the bark/bite thresholds (both in ms).
+    // If the bark threshold > bite threshold, the bite threshold is set to
+    // the bark threshold.
+    pub fn new(
+        base_addr: u32,
+        clock_freq_hz: u32,
+        bark_thold_ms: u64,
+        bite_thold_ms: u64,
+        bark_cb: Option<fn()>,
+    ) -> Result<Self, ()> {
+        let inst = AonTimer {
+            base_addr: slice_from_raw_parts_mut(base_addr as *mut u32, 12),
+            clock_freq_hz,
+            bark_cb,
+        };
+        unsafe {
+            // This is just a counter, reset it to 0.
+            (*inst.base_addr)[WDOG_COUNT_IDX] = 0;
+            // write 1 to clear any pending interrupts
+            (*inst.base_addr)[INTR_STATE_IDX] =
+                INTR_STATE_BITS_WKUP | INTR_STATE_BITS_WDOG;
```

Comment:
This whole driver is dealing with watchdog only. So, why does it enable the wakeup timer interrupt?

Reply:
So according to the [spec](https://docs.opentitan.org/hw/ip/aon_timer/doc/#Reg_intr_state), the code here is clearing the `INTR_STATE` register by writing 1 to each bit.

Reply:
Right. I meant that this should be `(*inst.base_addr)[INTR_STATE_IDX] = INTR_STATE_BITS_WDOG;` instead since `INTR_STATE_BITS_WKUP` is not really related to watchdog timer.

Reply:
If it's been decided already to have the driver solely focus on the watchdog part of the AON Timer IP, then I would go as far as to say to remove any reference to the wkup timer from this driver, since it's not needed and raises confusion.

Reply:
> If it's been decided already to have the driver solely focus on the watchdog part of the AON Timer IP, then I would go as far as to say to remove any reference to the wkup timer from this driver, since it's not needed and raises confusion.

That was exactly the motivation behind my question :) - to understand if this driver is supposed to deal with "watchdog" only or both "watchdog" and "wakeup". There are just 3 places where we reference wakeup timer here -
1. INTR_STATE_BITS_WKUP definition
2. clear_wdt_irq
3. new

And the comment on line 35 says "Initialize a new watchdog driver device". I am fine either ways. Just wanted to understand the plan here.

Reply:
Originally when it was written it was supposed to be both, but because we didn't know exactly what we were going to do with the wakeup timer (as it wasn't implemented yet and we have no way of testing it at the moment), it was punted to the todo-list. 

The demo app itself is concerned with the watchdog timer, but the driver could have both in the future, or we could have two separate drivers for each timer; their functions aren't related to each other. I'm not aware of any decision on that yet and I would imagine we'd want to wait until we know more about how the wakeup timer will work (or is implemented).

Reply:
In that case, would it make sense to just enable watchdog timer here and punt the addition of wakeup timer to later?

Reply:
Changed to just clear the watchdog interrupt bit.